### PR TITLE
Enable q_scale calibration for deepseek v2 on Gaudi

### PR DESCRIPTION
**Pull request by mengniwang95**
https://github.com/intel/auto-round/issues/1274

- Replace the deepseek v2 modeling to skip complex tensor issue on hpu
- Enable q_scale cache for deepseek v2
- Use original query rather than the qdq query for later inference

**How to use fp8 sdpa on hpu for deepseek v2**
```bash
auto-round --model /disk8/deepseek-ai/DeepSeek-V2-Lite-Chat/ --scheme FP8_STATIC --static_attention_dtype fp8 --iters 0 --format llm_compressor --disable_trust_remote_code
```

gsm8k result:
```bash
vllm (pretrained=/disk9/mengni/tmp_autoround/DeepSeek-V2-Lite-Chat-w8afp8/,tensor_parallel_size=4,enable_expert_parallel=True,max_model_len=4096,max_num_seqs=64,gpu_memory_utilization=0.85,dtype=bfloat16,max_gen_toks=2048,enable_prefix_caching=False,max_num_batched_tokens=32768,kv_cache_dtype=fp8_inc), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 128
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6581|±  |0.0131|
|     |       |strict-match    |     5|exact_match|↑  |0.6558|±  |0.0131|
```

### CODE REVIEW

File: auto_round/experimental/attention.py [hunk 1]
```diff
@@ -95,7 +95,7 @@ def forward(
             cur_query_max.detach().to(getattr(module, QUERY_MAX_NAME).data.device),
         )
         update_parameter_data(module, query_max, QUERY_MAX_NAME)
-        query, query_scale = per_tensor_fp8_qdq(query, tensor_max=query_max)
+        _, query_scale = per_tensor_fp8_qdq(query, tensor_max=query_max)
```
Comment (yiliu30) [inline auto_round/experimental/attention.py:98]: Do you see any accuracy improvement compared with the QDQ query?
If not, I’d prefer to keep the current behavior, or we can add a flag to control it.

File: auto_round/modelling/replace_modules.py [hunk 2]
```diff
@@ -28,6 +28,10 @@
 }
 
 
+if is_hpex_available():
```
Comment (yiliu30) [inline auto_round/modelling/replace_modules.py]: How about moving this check to `is_to_be_replaced`?

https://github.com/intel/auto-round/blob/e383eea3647110ffae4f84d04aa2958d97e4909d/auto_round/modelling/replace_modules.py#L94

File: auto_round/__main__.py [hunk 3]
```diff
@@ -159,6 +159,12 @@ def __init__(self, *args, **kwargs):
         basic.add_argument(
             "--enable_torch_compile", action="store_true", help="Enable PyTorch compilation for faster execution. "
         )
+        basic.add_argument(
+            "--disable_trust_remote_code",
```
Comment (yiliu30) [inline auto_round/__main__.py:163]: How about align with the transformers api to use `trust_remote_code`, the default value is `False`.

File: auto_round/special_model_handler.py [hunk 4]
```diff
@@ -52,10 +52,18 @@ def _handle_special_model(model):
     return model
 
 
-def update_module(model, formats: list[OutputFormat] = None):
+def update_module(model, formats: list[OutputFormat] = None, trust_remote_code: bool = True):
     if formats is not None and any([format_.is_gguf() for format_ in formats]):
         return model
 
+    # only update deepseek_v2 module when not trust_remote_code and on hpu
```
Comment (yiliu30) [inline auto_round/special_model_handler.py]: If we use `is_to_be_replaced`, this part can be removed?

File: auto_round/modelling/deepseek_v2.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+        self.config = original.config
+        self.rope_init_fn = original.rope_init_fn
+
+        self.attention_scaling = original.attention_scaling
+        self.register_buffer("inv_freq", original.original_inv_freq, persistent=False)
+        self.original_inv_freq = original.original_inv_freq
+
+    @torch.no_grad()
+    @dynamic_rope_update
```
Comment (yiliu30) [inline auto_round/modelling/deepseek_v2.py]: Why add this one?

File: auto_round/modelling/deepseek_v2.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        cls,
+        original: torch.nn.Module,
+        config: DeepseekV2Config,
+        **kwargs,
+    ) -> "DeepseekV2RotaryEmbedding":
+        """Create an instance from the original module."""
+        return cls(original, config)
+
+
+class DeepseekV2Attention(ReplacementModuleBase):
```
Comment (yiliu30) [inline auto_round/modelling/deepseek_v2.py:178]: Can we inherit from the original `DeepseekV2Attention` and only replace the ROPE part?

**Review by yiliu30 — COMMENTED** [review]
Nice work! left a few nit comments.

File: auto_round/experimental/attention.py (hunk 1 above)
Comment (mengniwang95) [reply auto_round/experimental/attention.py:98]: gsm8k is better after changed

File: auto_round/modelling/deepseek_v2.py (hunk 5 above)
Comment (mengniwang95) [reply auto_round/modelling/deepseek_v2.py]: keep same with transformers' code

File: auto_round/__main__.py (hunk 3 above)
Comment (mengniwang95) [reply auto_round/__main__.py:163]: Previously AR use trust_remote_code=True as the default value to load model, I am not sure the impact of changing it.

File: auto_round/experimental/attention.py (hunk 1 above)
Comment (yiliu30) [reply auto_round/experimental/attention.py:98]: I see. Would you mind aligning this behavior for K/V as well? It’s also fine to leave it as‑is if you’d prefer to merge it first

File: auto_round/__main__.py (hunk 3 above)
Comment (yiliu30) [reply auto_round/__main__.py:163]: @n1ck-guo  could you help double-check? Thanks!

**Review by yiliu30 — COMMENTED** [review]
We may add some UTs for that replacement; please check if this model works: https://huggingface.co/tflsxyy/DeepSeek-V3-bf16-4layers

File: auto_round/special_model_handler.py (hunk 4 above)
Comment (mengniwang95) [reply auto_round/special_model_handler.py]: `is_to_be_replaced` is module-level check, is it appropriate to put this model-level check there?

File: auto_round/experimental/attention.py (hunk 1 above)
Comment (mengniwang95) [reply auto_round/experimental/attention.py:98]: Update the behavior of K/V, please check

File: auto_round/modelling/replace_modules.py (hunk 2 above)
Comment (mengniwang95) [reply auto_round/modelling/replace_modules.py]: removed

File: auto_round/modelling/deepseek_v2.py (hunk 6 above)
Comment (mengniwang95) [reply auto_round/modelling/deepseek_v2.py:178]: inherit DeepseekV2Attention and DeepseekV2RotaryEmbedding from original modules, their forward functions also need replacement

Comment (mengniwang95) [reply auto_round/modelling/deepseek_v2.py:178]: update replace way

File: auto_round/special_model_handler.py (hunk 4 above)
Comment (mengniwang95) [reply auto_round/special_model_handler.py]: updated

**Comment by mengniwang95** [comment]
> We may add some UTs for that replacement; please check if this model works: https://huggingface.co/tflsxyy/DeepSeek-V3-bf16-4layers

deepseev3 does not require this replacement, you can use --disable_trust_remote_code to catch the q_scale

File: auto_round/experimental/attention.py (hunk 1 above)
Comment (mengniwang95) [reply auto_round/experimental/attention.py:98]: Q qdq, kv no qdq
vllm (pretrained=/disk9/mengni/auto-round/tmp_autoround/DeepSeek-V2-Lite-Chat-w8afp8/,tensor_parallel_size=4,enable_expert_parallel=True,max_model_len=4096,max_num_seqs=64,gpu_memory_utilization=0.85,dtype=bfloat16,max_gen_toks=2048,enable_prefix_caching=False,max_num_batched_tokens=32768,kv_cache_dtype=fp8_inc), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 128
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6558|±  |0.0131|
|     |       |strict-match    |     5|exact_match|↑  |0.6528|±  |0.0131|


Q no qdq, kv no qdq
vllm (pretrained=/disk9/mengni/auto-round/tmp_autoround/DeepSeek-V2-Lite-Chat-w8afp8/,tensor_parallel_size=4,enable_expert_parallel=True,max_model_len=4096,max_num_seqs=64,gpu_memory_utilization=0.85,dtype=bfloat16,max_gen_toks=2048,enable_prefix_caching=False,max_num_batched_tokens=32768,kv_cache_dtype=fp8_inc), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 128
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.652|±  |0.0131|
|     |       |strict-match    |     5|exact_match|↑  |0.652|±  |0.0131|


Q no qdq, kv  qdq
vllm (pretrained=/disk9/mengni/auto-round/tmp_autoround/DeepSeek-V2-Lite-Chat-w8afp8/,tensor_parallel_size=4,enable_expert_parallel=True,max_model_len=4096,max_num_seqs=64,gpu_memory_utilization=0.85,dtype=bfloat16,max_gen_toks=2048,enable_prefix_caching=False,max_num_batched_tokens=32768,kv_cache_dtype=fp8_inc), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 128
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6581|±  |0.0131|
|     |       |strict-match    |     5|exact_match|↑  |0.6558|±  |0.0131|

Q qdq, kv  qdq
vllm (pretrained=/disk9/mengni/auto-round/tmp_autoround/DeepSeek-V2-Lite-Chat-w8afp8/,tensor_parallel_size=4,enable_expert_parallel=True,max_model_len=4096,max_num_seqs=64,gpu_memory_utilization=0.85,dtype=bfloat16,max_gen_toks=2048,enable_prefix_caching=False,max_num_batched_tokens=32768,kv_cache_dtype=fp8_inc), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 128
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6467|±  |0.0132|
|     |       |strict-match    |     5|exact_match|↑  |0.6459|±  |0.0132|

Comment (yiliu30) [reply auto_round/experimental/attention.py:98]: Great, really solid work!

File: auto_round/modelling/deepseek_v2.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        """Create an instance from the original module."""
+        original.forward = partial(rotary_emb_forward, original)
+        return original
+
+    @classmethod
+    def is_to_be_replaced(
+        cls,
+        original: torch.nn.Module,
+    ) -> bool:
+        """Determine if the given module should be replaced.
```
Comment (yiliu30) [inline auto_round/modelling/deepseek_v2.py]: Please update docstring, thanks!

Comment (mengniwang95) [reply auto_round/modelling/deepseek_v2.py]: Thanks, updated

File: auto_round/__main__.py (hunk 3 above)
Comment (mengniwang95) [reply auto_round/__main__.py:163]: Merge this PR first, check this flag later
